File: lending-database/src/main/java/info/armado/ausleihe/database/access/Dao.java
/**
 * 
 */
package info.armado.ausleihe.database.access;

import java.io.Serializable;
import java.util.List;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import javax.transaction.Transactional;

import lombok.NoArgsConstructor;

/**
 * @author Arndt
 *
 */
@NoArgsConstructor
public abstract class Dao<Entity, PK> implements Serializable {
	/**
	 * 
	 */
	private static final long serialVersionUID = 1L;

	protected Class<Entity> type;

	@PersistenceContext(unitName = "ausleihe")
	protected EntityManager em;

	protected Dao(Class<Entity> type) {
		this.type = type;
	}

	/**
	 * This method persists a list of entities in a single transaction. This
	 * means that if one of the given entities can not be persisted the call
	 * fails and all already persisted entities from this method calls get
	 * rerolled
	 * 
	 * @param games
	 *            The entities to be persisted in the database
	 */
	@Transactional
	public void insert(List<Entity> entities) {
		for (Entity entity : entities) {
			this.insert(entity);
		}
	}

	/**
	 * This method persists a given object of type Entity.
	 * 
	 * @param entity
	 *            The entity to be persisted
	 */
	public void insert(Entity entity) {
		em.persist(entity);
	}

	/**
	 * This method removes/deletes a given object of type Entity.
	 * 
	 * @param entity
	 *            The entity to be removed
	 */
	public void delete(Entity entity) {
		em.remove(entity);
	}

	/**
	 * This method updates a given object of type Entity.
	 * 
	 * @param entity
	 *            The entity to be updated
	 */
	@Transactional
	public void update(Entity entity) {
		em.merge(entity);
	}

	/**
	 * This method refreshes the given object of type Entity. In this process
	 * its state may be overridden.
	 * 
	 * @param entity
	 *            The entity to be refreshed
	 */
	public void refresh(Entity entity) {
		em.refresh(entity);
	}

	/**
	 * This method returns the entity corresponding to the given primary key.
	 * 
	 * @param pk
	 *            The primary key from the searched object of type Entity
	 * @return The searched object with the given primary key or null if no such
	 *         object exists
	 */
	public Entity select(PK pk) {
		return em.find(type, pk);
	}

	/**
	 * This method returns all persisted objects of type Entity.
	 * 
	 * @return A list of all persisted objects of type Entity
	 */
	public List<Entity> selectAll() {
		return em.createQuery("from " + type.getSimpleName(), type).getResultList();
	}
	
	/**
	 * This method deletes all persisted objects of type Entity.
	 */
	@Transactional
	public void deleteAll() {
		em.createQuery(String.format("delete from %s", type.getSimpleName())).executeUpdate();
	}
}

File: lending-database/src/main/scala/info/armado/ausleihe/database/access/Dao.scala
package info.armado.ausleihe.database.access

import javax.persistence.{EntityManager, PersistenceContext}
import javax.transaction.Transactional

import scala.collection.JavaConverters.asScalaBufferConverter
import scala.collection.TraversableOnce
import scala.util.Try

/**
  * An abstract Data Access Object, which contains all basic functionality needed to work with an entity type [[Entity]].
  *
  * @param entityType The class of the entity [[Entity]]
  * @tparam Entity The type of entity
  * @tparam PK     The type of the primary key of the entity
  * @author Marc Arndt
  * @since 25.06.17
  */
abstract class Dao[Entity, PK](protected val entityType: Class[Entity]) extends Serializable {
  /**
    * The [[EntityManager]] for this [[Dao]].
    * This [[EntityManager]] will be injected by the application server
    */
  @PersistenceContext(unitName = "ausleihe")
  protected var em: EntityManager = _

  /**
    * Inserts a given [[Entity]] in the database
    *
    * @param entity The entity to be persisted/inserted
    */
  @Transactional
  def insert(entity: Entity): Unit = em.persist(entity)

  /**
    * Inserts a collection of entities in the database
    *
    * @param entities The collection of entities to be persisted/inserted
    */
  @Transactional
  def insert(entities: TraversableOnce[Entity]): Unit = entities.foreach(insert)

  /**
    * Deletes a given [[Entity]] from the database
    *
    * @param entity The entity to be removed/deleted
    */
  @Transactional
  def delete(entity: Entity): Unit = em.remove(entity)

  /**
    * Deletes a collection of entities from the database
    *
    * @param entities The collection of entities to be removed/deleted
    */
  @Transactional
  def delete(entities: TraversableOnce[Entity]): Unit = entities.foreach(delete)

  /**
    * Updates a given [[Entity]] in the database
    *
    * @param entity The to be updated entity with its new and updated values
    */
  @Transactional
  def update(entity: Entity): Unit = {
    em.merge(entity)
    /*
     * TODO: remove the flush() operation at a later time
     * The flush is a quick fix to have @Transactional operate correctly during the admin server tests
     */
    em.flush()
  }

  /**
    * Updates a collection of entities in the database
    *
    * @param entities The collection of entities to be updated
    */
  @Transactional
  def update(entities: TraversableOnce[Entity]): Unit = entities.foreach(update)

  /**
    * Refreshes a given [[Entity]] instance with the current state taken from the database
    *
    * @param entity The entity to be refreshed
    */
  def refresh(entity: Entity): Unit = em.refresh(entity)

  /**
    * Refreshes a collection of entities with the current state taken from the database
    *
    * @param entities The collection of entities to be refreshed
    */
  def refresh(entities: TraversableOnce[Entity]): Unit = entities.foreach(refresh)

  /**
    * Queries the database for an [[Entity]] with the given primary key
    *
    * @param pk The primary key of the queried entity
    * @return The found entity wrapped inside a [[Some]] instance or [[None]] if no such entity could be found
    */
  def select(pk: PK): Option[Entity] = Try(em.find(entityType, pk)).toOption

  /**
    * Queries a list of all entities of type `entityType` from the database
    *
    * @return A list containing all entities of type `entityType`
    */
  def selectAll(): List[Entity] =
    em.createQuery(s"from ${entityType.getSimpleName}", entityType).getResultList.asScala.toList

  /**
    * Clears the whole table containing the `entityType` inside the database
    */
  @Transactional
  def deleteAll(): Unit = em.createQuery(s"delete from ${entityType.getSimpleName}").executeUpdate
}
